extensions/azurecore/src/account-provider/azureAccountProvider2.ts

```diff
@@ -0,0 +1,443 @@
+		return this._initialize(storedAccounts);
+	}
+
+	private async _initialize(storedAccounts: azdata.Account[]): Promise<azdata.Account[]> {
+		for (let account of storedAccounts) {
+			try {
+				const x = await this.getAccessTokens(account, azdata.AzureResource.ResourceManagement);
+				console.log(x);
+			} catch (e) {
+				console.error(`Refreshing account ${account.displayInfo} failed - ${e}`);
+				account.isStale = true;
+				azdata.accounts.accountUpdated(account);
+			}
+		}
+		this.isInitialized = true;
+		return storedAccounts;
+	}
+
+	private async getToken(userId: string, tenantId: string, resourceId: string): Promise<TokenResponse> {
+		let authorityUrl = url.resolve(this.metadata.settings.host, tenantId);
+		const context = new AuthenticationContext(authorityUrl, null, this._tokenCache);
+
+		const acquireToken = promisify(context.acquireToken).bind(context);
+
+		let response: (TokenResponse | ErrorResponse) = await acquireToken(resourceId, userId, this.metadata.settings.clientId);
+		if (response.error) {
+			throw new Error(`Response contained error ${response}`);
+		}
+
+		response = response as TokenResponse;
+
+		context.cache.add([response], (err, result) => {
+			console.log(err, result);
+		});
+
+		return response;
+	}
+
+	private async getAccessTokens(account: azdata.Account, resource: azdata.AzureResource): Promise<AzureAccountSecurityTokenCollection> {
+		const resourceIdMap = new Map<azdata.AzureResource, string>([
+			[azdata.AzureResource.ResourceManagement, this.metadata.settings.armResource.id],
+			[azdata.AzureResource.Sql, this.metadata.settings.sqlResource.id]
+		]);
+		const tenantRefreshPromises: Promise<{ tenantId: any, securityToken: AzureAccountSecurityToken }>[] = [];
+		const tokenCollection: AzureAccountSecurityTokenCollection = {};
+
+		for (let tenant of account.properties.tenants) {
+			const promise = new Promise<{ tenantId: any, securityToken: AzureAccountSecurityToken }>(async (resolve, reject) => {
+				try {
+					let response = await this.getToken(tenant.userId, tenant.id, resourceIdMap.get(resource));
+
+					resolve({
+						tenantId: tenant.id,
+						securityToken: {
+							expiresOn: response.expiresOn,
+							resource: response.resource,
+							token: response.accessToken,
+							tokenType: response.tokenType
+						} as AzureAccountSecurityToken,
+					});
+				} catch (ex) {
+					reject(ex);
+				}
+
+			});
+			tenantRefreshPromises.push(promise);
+		}
+
+		const refreshedTenants = await Promise.all(tenantRefreshPromises);
+		refreshedTenants.forEach((refreshed) => {
+			tokenCollection[refreshed.tenantId] = refreshed.securityToken;
+		});
+
+		return tokenCollection;
+	}
+
+	// interface method
+	getSecurityToken(account: azdata.Account, resource: azdata.AzureResource): Thenable<{}> {
+		return this._getSecurityToken(account, resource);
+	}
+
+	private async _getSecurityToken(account: azdata.Account, resource: azdata.AzureResource): Promise<{}> {
```

Comment:
Doesn't really seem necessary to wrap this in another function - you could just have getSecurityToken call getAccessTokens directly

Reply:
Ah I did this for all of them when I just started the new implementation because I wanted to be able to use Promise and async together.

I think I may just keep it just to make it fit the whole design of the other methods.

Good catch though!